cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java

```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
 
     service.setSpec(spec);
 
-    try {
-      api.createNamespacedService(namespace, service, null, null, null);
-      LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
-    } catch (ApiException e) {
-      // It means the service already exists. In this case we update the port if it is not the same.
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
-        return false;
+    // Maximum elapse-able time currently set to 10 minutes
+    ExponentialBackOff backoffGenerator
+        = new ExponentialBackOff.Builder().setMaxElapsedTimeMillis(600000).build();
+    boolean serviceCreatedSuccessfully = false;
+    long backoffDuration = ExponentialBackOff.DEFAULT_INITIAL_INTERVAL_MILLIS;
+
+    while (!serviceCreatedSuccessfully) {
+      try {
+        api.createNamespacedService(namespace, service, null, null, null);
+        LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
+        serviceCreatedSuccessfully = true;
+      } catch (ApiException e) {
+        // It means the service already exists. In this case we update the port if it is not the same.
+        if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
+          return false;
+        } else {
+          if (e.getCause() instanceof SocketTimeoutException) {
+            LOG.warn("Service creation in kubernetes with name {} and port {} encountered exception: {}",
+                     serviceName,
+                     port.getPort(),
+                     e);
+            try {
+              Thread.sleep(backoffDuration);
```

Comment:
I am a bit concerned that we block the caller thread. Typically service registration shouldn't be blocked like this. It should either fail fast (like the current behavior), or retry in the background (which is kind of more complicated to handle).